File: leetcode/linked_list_cycleII.js
/* Given a linked list, return the node where the cycle begins. 
 * If there is no cycle, return null.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    for (var slow = head, fast = head; fast; ) {
        slow = slow.next;
        fast = fast.next;
        if (!fast)
            break;
        fast = fast.next;
        if (slow === fast)
            break;
    }
    if (!fast)
        return null;
    while (head !== slow) {
        head = head.next;
        slow = slow.next;
    }
    return head;
};

File: lintcode/linked_list_cycleII.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: The node where the cycle begins. 
                if there is no cycle, return null
    """
    def detectCycle(self, head):
        slow = fast = head
        while fast:
            slow, fast = slow.next, fast.next
            if not fast: break
            fast = fast.next
            if slow == fast: break
        if not fast: return None
        while head != slow:
            head, slow = head.next, slow.next
        return head
